exp/internal/controllers/machinepool_controller_phases.go

```diff
@@ -317,3 +324,259 @@ func (r *MachinePoolReconciler) reconcileInfrastructure(ctx context.Context, clu
+	machineList := &clusterv1.MachineList{}
+	labels := map[string]string{
+		expv1.MachinePoolNameLabel:      mp.Name,
+		clusterv1.ClusterNameLabel:      mp.Spec.ClusterName,
+		clusterv1.MachinePoolOwnedLabel: "true",
+	}
+
+	if err := r.Client.List(ctx, machineList, client.InNamespace(mp.Namespace), client.MatchingLabels(labels)); err != nil {
+		return err
+	}
+
+	for i := range machineList.Items {
+		machine := &machineList.Items[i]
+		infraRef := machine.Spec.InfrastructureRef
+
+		// Check this here since external.Get() will complain if the infraRef is empty, i.e. no object kind or apiVersion.
+		if (infraRef == corev1.ObjectReference{}) {
+			log.V(2).Info("Machine has an empty infraRef, this is likely a fallback machine, will delete anyway", "machine", machine.Name, "namespace", machine.Namespace)
+
+			return r.Client.Delete(ctx, machine)
+		}
+
+		_, err := external.Get(ctx, r.Client, &infraRef, mp.Namespace)
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				log.Error(err, "unexpected error while trying to get infra machine, will delete anyway", "machine", machine.Name, "namespace", machine.Namespace)
+			}
+
+			log.V(2).Info("Deleting orphaned machine", "machine", machine.Name, "namespace", machine.Namespace)
+			if err := r.Client.Delete(ctx, machine); err != nil {
+				return err
+			}
+		} else {
+			log.V(2).Info("Machine is not orphaned, nothing to do", "machine", machine.Name, "namespace", machine.Namespace)
+		}
+	}
+
+	return nil
+}
+
+// createMachinesIfNotExists creates a MachinePool Machine for each infraMachine if it doesn't already exist and sets the owner reference and infraRef.
+func (r *MachinePoolReconciler) createMachinesIfNotExists(ctx context.Context, mp *expv1.MachinePool, infraMachines []unstructured.Unstructured) error {
+	machineList := &clusterv1.MachineList{}
+	labels := map[string]string{
+		expv1.MachinePoolNameLabel:      mp.Name,
+		clusterv1.ClusterNameLabel:      mp.Spec.ClusterName,
+		clusterv1.MachinePoolOwnedLabel: "true",
+	}
+
+	if err := r.Client.List(ctx, machineList, client.InNamespace(mp.Namespace), client.MatchingLabels(labels)); err != nil {
+		return err
+	}
+
+	setOfInfraRefNames := map[string]struct{}{}
+	for _, machine := range machineList.Items {
+		infraRef := machine.Spec.InfrastructureRef
+		setOfInfraRefNames[infraRef.Name] = struct{}{}
+	}
+
+	for i := range infraMachines {
+		infraMachine := &infraMachines[i]
+		if _, infraRefExists := setOfInfraRefNames[infraMachine.GetName()]; infraRefExists {
+			continue
+		}
+
+		machine := getNewMachine(mp, infraMachine, false)
+		if err := r.Client.Create(ctx, machine); err != nil {
+			return errors.Wrapf(err, "failed to create new Machine for infraMachine %q in namespace %q", infraMachine.GetName(), infraMachine.GetNamespace())
+		}
+
+		// Set the owner reference on the infraMachine to the Machine since the infraMachine is created and owned by the infraMachinePool.
+		infraMachine.SetOwnerReferences([]metav1.OwnerReference{
```

Comment:
why not set the owner ref on the machine directly before creating it instead of creating is and then having to patch it separately?

Reply:
This is for patching the infraMachine so that it's owned by the Machine. When the infraMachine is created, it's corresponding machine doesn't exist yet so we can't set the owner at that time.

Reply:
Got it, I missed that that was patching the infraMachine, not the machine 👍